README.md
# ![Devices by Year Class](https://github.com/facebook/device-year-class/raw/master/docs/images/logo_trans_square.png) Device Year Class

Device Year Class is an Android library that implements a simple algorithm that maps
a device's RAM, CPU cores, and clock speed to the year where those combination
of specs were considered high end. This allows a developer to easily modify
application behavior based on the capabilities of the phone's hardware.

![Most Popular Devices by Year Class](https://github.com/facebook/device-year-class/raw/master/docs/images/popular_devices_by_year_class.png)

Mappings as of this writing (RAM is a ceiling):

| RAM | condition | Year Class |
|----:|----------:|-----------:|
|768MB| 1 core    | 2009 |
|     | 2+ cores  | 2010 |
|  1GB| <1.3GHz   | 2011 |
|     | 1.3GHz+   | 2012 |
|1.5GB| <1.8GHz   | 2012 |
|     | 1.8GHz+   | 2013 |
|  2GB|           | 2013 |
|  3GB|           | 2014 |
|  5GB|           | 2015 |
| more|           | 2016 |

## Integration

### Download
Download [the latest JARs](https://github.com/facebook/device-year-class/releases/latest) or grab via Gradle:
```groovy
compile 'com.facebook.device.yearclass:yearclass:2.1.0'
```
or Maven:
```xml
<dependency>
  <groupId>com.facebook.device.yearclass</groupId>
  <artifactId>yearclass</artifactId>
  <version>2.1.0</version>
</dependency>
```

### Calculate Device Year Class
Calculating the current device's Year Class is simple.

```java
int year = YearClass.get(getApplicationContext());
```

Then, later on, you can use the year class to make decisions in your app, or
send it along with your analytics.

```java
if (year >= 2013) {
    // Do advanced animation
} else if (year > 2010) {
    // Do simple animation
} else {
    // Phone too slow, don't do any animations
}
```

See the `yearclass-sample` project for more details.

## Improve Device Year Class!
See the [CONTRIBUTING.md](https://github.com/facebook/device-year-class/blob/master/CONTRIBUTING.md) file for how to help out.

## License
Device Year Class is [BSD-licensed](https://github.com/facebook/device-year-class/blob/master/LICENSE). We also provide an additional patent grant.


CHANGELOG.md
Change Log
==========

## Version 2.1.0

 * Added Year Class 2016 to classification model.
 
## Version 2.0.0

_2018-04-13_

 * New classification model.

## Version 1.0.1

_2015-04-02_

 * Threading fix.


## Version 1.0.0

_2015-03-25_

 * Initial release.


yearclass/src/test/java/com/facebook/device/yearclass/YearClassTest.java
/* Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.facebook.device.yearclass;

import android.content.Context;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import static org.junit.Assert.*;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.*;

@RunWith(PowerMockRunner.class)
public class YearClassTest {

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testGetYearCategoryS7() {
    // CPU, frequency, RAM, and YearClass values from Samsung Galaxy S7 (global edition).
    int yearClass = getYearClass(8, 2600000, 3663L * 1024 * 1024);
    assertEquals(YearClass.CLASS_2015, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testGetYearCategoryG4() {
    // CPU, frequency, RAM, and YearClass values from LG G4.
    int yearClass = getYearClass(6, 1824000, 2778L * 1024 * 1024);
    assertEquals(YearClass.CLASS_2014, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testGetYearCategoryS5() {
    // CPU, frequency, RAM, and YearClass values from Samsung Galaxy S5.
    int yearClass = getYearClass(4, 2457600, 1946939392L);
    assertEquals(YearClass.CLASS_2013, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testGetYearCategoryGalaxyJ1() {
    // CPU, frequency, RAM, and YearClass values from Samsung Galaxy J1.
    int yearClass = getYearClass(2, 1248000, 716L * 1024 * 1024);
    assertEquals(YearClass.CLASS_2010, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testGetYearCategoryP8lite() {
    // CPU, frequency, RAM, and YearClass values from Huawei P8lite.
    int yearClass = getYearClass(8, 1200000, 1858L * 1024 * 1024);
    assertEquals(YearClass.CLASS_2013, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testEmptyCase() {
    int yearClass = getYearClass(DeviceInfo.DEVICEINFO_UNKNOWN,
        DeviceInfo.DEVICEINFO_UNKNOWN, DeviceInfo.DEVICEINFO_UNKNOWN);
    assertEquals(YearClass.CLASS_UNKNOWN, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testCoreNums() {
    //Test with only number of cores information available.
    int yearClass = getYearClass(4,
        DeviceInfo.DEVICEINFO_UNKNOWN, DeviceInfo.DEVICEINFO_UNKNOWN);
    assertEquals(YearClass.CLASS_2012, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testClockSpeed() {
    //Test with only clock speed information available.
    int yearClass = getYearClass(DeviceInfo.DEVICEINFO_UNKNOWN,
        2457600, DeviceInfo.DEVICEINFO_UNKNOWN);
    assertEquals(YearClass.CLASS_2014, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testTotalRAM() {
    //Test with only total RAM information available.
    int yearClass = getYearClass(DeviceInfo.DEVICEINFO_UNKNOWN,
        DeviceInfo.DEVICEINFO_UNKNOWN, 1946939392L);
    assertEquals(YearClass.CLASS_2013, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testGetYearCategoryS7() {
    // CPU, frequency, RAM, and YearClass values from Samsung Galaxy Note8 (USA edition).
    int yearClass = getYearClass(8, 2350000, 6144L * 1024 * 1024);
    assertEquals(YearClass.CLASS_2016, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testTotalRAM() {
    //Test with only total RAM information available.
    int yearClass = getYearClass(DeviceInfo.DEVICEINFO_UNKNOWN,
        DeviceInfo.DEVICEINFO_UNKNOWN, 6144L * 1024 * 1024);
    assertEquals(YearClass.CLASS_2016, yearClass);
  }

  private int getYearClass(int numCores, int maxFreqKHz, long memoryBytes) {
    mockStatic(DeviceInfo.class);
    when(DeviceInfo.getNumberOfCPUCores()).thenReturn(numCores);
    when(DeviceInfo.getCPUMaxFreqKHz()).thenReturn(maxFreqKHz);
    when(DeviceInfo.getTotalMemory((Context) any())).thenReturn(memoryBytes);
    int yearClass = YearClass.get(null);
    PowerMockito.verifyStatic();

    return yearClass;
  }
}


yearclass/src/test/java/com/facebook/device/yearclass/DeviceInfoTest.java
/* Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.facebook.device.yearclass;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class DeviceInfoTest {

    @Test
    public void testFileStringValid() {
        assertEquals(DeviceInfo.getCoresFromFileString("0-3"), 4);
        assertEquals(DeviceInfo.getCoresFromFileString("0-11"), 12);
    }

    @Test
    public void testFileStringInvalid() {
        assertEquals(DeviceInfo.getCoresFromFileString("INVALIDSTRING"), -1);
        assertEquals(DeviceInfo.getCoresFromFileString("0-2a"), -1);
        assertEquals(DeviceInfo.getCoresFromFileString("035"), -1);
    }
}


yearclass-sample/src/test/java/com/facebook/device/yearclass/sample/MainActivityTest.java
package com.facebook.device.yearclass.sample;

import com.facebook.device.yearclass.YearClass;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;

@RunWith(RobolectricTestRunner.class)
public class MainActivityTest {
	@Test
	public void SampleTest() {
        MainActivity activity = Robolectric.setupActivity(MainActivity.class);
        assertEquals(YearClass.get(activity), 1);
	}

}


yearclass/src/main/java/com/facebook/device/yearclass/DeviceInfo.java
/* Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.facebook.device.yearclass;

import android.annotation.TargetApi;
import android.app.ActivityManager;
import android.content.Context;
import android.os.Build;

import java.io.*;
/**
 * Helper class for accessing hardware specifications, including the number of CPU cores, CPU clock speed
 * and total available RAM.
 */
public class DeviceInfo {

  /**
   * The default return value of any method in this class when an
   * error occurs or when processing fails (Currently set to -1). Use this to check if
   * the information about the device in question was successfully obtained.
   */
  public static final int DEVICEINFO_UNKNOWN = -1;

  /**
   * Reads the number of CPU cores from the first available information from
   * {@code /sys/devices/system/cpu/possible}, {@code /sys/devices/system/cpu/present},
   * then {@code /sys/devices/system/cpu/}.
   *
   * @return Number of CPU cores in the phone, or DEVICEINFO_UKNOWN = -1 in the event of an error.
   */
  public static int getNumberOfCPUCores() {
    if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.GINGERBREAD_MR1) {
      // Gingerbread doesn't support giving a single application access to both cores, but a
      // handful of devices (Atrix 4G and Droid X2 for example) were released with a dual-core
      // chipset and Gingerbread; that can let an app in the background run without impacting
      // the foreground application. But for our purposes, it makes them single core.
      return 1;
    }
    int cores;
    try {
      cores = getCoresFromFileInfo("/sys/devices/system/cpu/possible");
      if (cores == DEVICEINFO_UNKNOWN) {
        cores = getCoresFromFileInfo("/sys/devices/system/cpu/present");
      }
      if (cores == DEVICEINFO_UNKNOWN) {
        cores = getCoresFromCPUFileList();
      }
    } catch (SecurityException e) {
      cores = DEVICEINFO_UNKNOWN;
    } catch (NullPointerException e) {
      cores = DEVICEINFO_UNKNOWN;
    }
    return cores;
  }

  /**
   * Tries to read file contents from the file location to determine the number of cores on device.
   * @param fileLocation The location of the file with CPU information
   * @return Number of CPU cores in the phone, or DEVICEINFO_UKNOWN = -1 in the event of an error.
   */
  private static int getCoresFromFileInfo(String fileLocation) {
    InputStream is = null;
    try {
      is = new FileInputStream(fileLocation);
      BufferedReader buf = new BufferedReader(new InputStreamReader(is));
      String fileContents = buf.readLine();
      buf.close();
      return getCoresFromFileString(fileContents);
    } catch (IOException e) {
      return DEVICEINFO_UNKNOWN;
    } finally {
      if (is != null) {
        try {
          is.close();
        } catch (IOException e) {
            // Do nothing.
        }
      }
    }
  }

  /**
   * Converts from a CPU core information format to number of cores.
   * @param str The CPU core information string, in the format of "0-N"
   * @return The number of cores represented by this string
   */
  static int getCoresFromFileString(String str) {
    if (str == null || !str.matches("0-[\\d]+$")) {
      return DEVICEINFO_UNKNOWN;
    }
    int cores = Integer.valueOf(str.substring(2)) + 1;
    return cores;
  }

  private static int getCoresFromCPUFileList() {
    return new File("/sys/devices/system/cpu/").listFiles(CPU_FILTER).length;
  }

  private static final FileFilter CPU_FILTER = new FileFilter() {
    @Override
    public boolean accept(File pathname) {
      String path = pathname.getName();
      //regex is slow, so checking char by char.
      if (path.startsWith("cpu")) {
        for (int i = 3; i < path.length(); i++) {
          if (!Character.isDigit(path.charAt(i))) {
            return false;
          }
        }
        return true;
      }
      return false;
    }
  };

  /**
   * Method for reading the clock speed of a CPU core on the device. Will read from either
   * {@code /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq} or {@code /proc/cpuinfo}.
   *
   * @return Clock speed of a core on the device, or -1 in the event of an error.
   */
  public static int getCPUMaxFreqKHz() {
    int maxFreq = DEVICEINFO_UNKNOWN;
    try {
      for (int i = 0; i < getNumberOfCPUCores(); i++) {
        String filename =
            "/sys/devices/system/cpu/cpu" + i + "/cpufreq/cpuinfo_max_freq";
        File cpuInfoMaxFreqFile = new File(filename);
        if (cpuInfoMaxFreqFile.exists() && cpuInfoMaxFreqFile.canRead()) {
          byte[] buffer = new byte[128];
          FileInputStream stream = new FileInputStream(cpuInfoMaxFreqFile);
          try {
            stream.read(buffer);
            int endIndex = 0;
            //Trim the first number out of the byte buffer.
            while (Character.isDigit(buffer[endIndex]) && endIndex < buffer.length) {
              endIndex++;
            }
            String str = new String(buffer, 0, endIndex);
            Integer freqBound = Integer.parseInt(str);
            if (freqBound > maxFreq) {
              maxFreq = freqBound;
            }
          } catch (NumberFormatException e) {
            //Fall through and use /proc/cpuinfo.
          } finally {
            stream.close();
          }
        }
      }
      if (maxFreq == DEVICEINFO_UNKNOWN) {
        FileInputStream stream = new FileInputStream("/proc/cpuinfo");
        try {
          int freqBound = parseFileForValue("cpu MHz", stream);
          freqBound *= 1000; //MHz -> kHz
          if (freqBound > maxFreq) maxFreq = freqBound;
        } finally {
          stream.close();
        }
      }
    } catch (IOException e) {
      maxFreq = DEVICEINFO_UNKNOWN; //Fall through and return unknown.
    }
    return maxFreq;
  }

  /**
   * Calculates the total RAM of the device through Android API or /proc/meminfo.
   *
   * @param c - Context object for current running activity.
   * @return Total RAM that the device has, or DEVICEINFO_UNKNOWN = -1 in the event of an error.
   */
  @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
  public static long getTotalMemory(Context c) {
    // memInfo.totalMem not supported in pre-Jelly Bean APIs.
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
      ActivityManager.MemoryInfo memInfo = new ActivityManager.MemoryInfo();
      ActivityManager am = (ActivityManager) c.getSystemService(Context.ACTIVITY_SERVICE);
      am.getMemoryInfo(memInfo);
      if (memInfo != null) {
        return memInfo.totalMem;
      } else {
        return DEVICEINFO_UNKNOWN;
      }
    } else {
      long totalMem = DEVICEINFO_UNKNOWN;
      try {
        FileInputStream stream = new FileInputStream("/proc/meminfo");
        try {
          totalMem = parseFileForValue("MemTotal", stream);
          totalMem *= 1024;
        } finally {
          stream.close();
        }
      } catch (IOException e) {
      }
      return totalMem;
    }
  }

  /**
   * Helper method for reading values from system files, using a minimised buffer.
   *
   * @param textToMatch - Text in the system files to read for.
   * @param stream      - FileInputStream of the system file being read from.
   * @return A numerical value following textToMatch in specified the system file.
   * -1 in the event of a failure.
   */
  private static int parseFileForValue(String textToMatch, FileInputStream stream) {
    byte[] buffer = new byte[1024];
    try {
      int length = stream.read(buffer);
      for (int i = 0; i < length; i++) {
        if (buffer[i] == '\n' || i == 0) {
          if (buffer[i] == '\n') i++;
          for (int j = i; j < length; j++) {
            int textIndex = j - i;
            //Text doesn't match query at some point.
            if (buffer[j] != textToMatch.charAt(textIndex)) {
              break;
            }
            //Text matches query here.
            if (textIndex == textToMatch.length() - 1) {
              return extractValue(buffer, j);
            }
          }
        }
      }
    } catch (IOException e) {
      //Ignore any exceptions and fall through to return unknown value.
    } catch (NumberFormatException e) {
    }
    return DEVICEINFO_UNKNOWN;
  }

  /**
   * Helper method used by {@link #parseFileForValue(String, FileInputStream) parseFileForValue}. Parses
   * the next available number after the match in the file being read and returns it as an integer.
   * @param index - The index in the buffer array to begin looking.
   * @return The next number on that line in the buffer, returned as an int. Returns
   * DEVICEINFO_UNKNOWN = -1 in the event that no more numbers exist on the same line.
   */
  private static int extractValue(byte[] buffer, int index) {
    while (index < buffer.length && buffer[index] != '\n') {
      if (Character.isDigit(buffer[index])) {
        int start = index;
        index++;
        while (index < buffer.length && Character.isDigit(buffer[index])) {
          index++;
        }
        String str = new String(buffer, 0, start, index - start);
        return Integer.parseInt(str);
      }
      index++;
    }
    return DEVICEINFO_UNKNOWN;
  }
}


yearclass/src/androidTest/java/com/facebook/device/yearclass/YearClassTest.java
/* Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.facebook.device.yearclass;

import android.content.Context;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import static org.junit.Assert.*;
import static org.mockito.Matchers.anyObject;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.*;

@RunWith(PowerMockRunner.class)
public class YearClassTest {

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testGetYearCategory() {
    // CPU, frequency, RAM, and YearClass values from Samsung Galaxy S5.
    YearClass yearClass = getYearClass(4, 2457600, 1946939392L);
    assertEquals(YearClass.CLASS_2013, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testEmptyCase() {
    YearClass yearClass = getYearClass(DeviceInfo.DEVICEINFO_UNKNOWN,
        DeviceInfo.DEVICEINFO_UNKNOWN, DeviceInfo.DEVICEINFO_UNKNOWN);
    assertEquals(YearClass.CLASS_UNKNOWN, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testCoreNums() {
    //Test with only number of cores information available.
    YearClass yearClass = getYearClass(4,
        DeviceInfo.DEVICEINFO_UNKNOWN, DeviceInfo.DEVICEINFO_UNKNOWN);
    assertEquals(YearClass.CLASS_2012, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testClockSpeed() {
    //Test with only clock speed information available.
    YearClass yearClass = getYearClass(DeviceInfo.DEVICEINFO_UNKNOWN,
        2457600, DeviceInfo.DEVICEINFO_UNKNOWN);
    assertEquals(YearClass.CLASS_2014, yearClass);
  }

  @PrepareForTest(DeviceInfo.class)
  @Test
  public void testTotalRAM() {
    //Test with only total RAM information available.
    YearClass yearClass = getYearClass(DeviceInfo.DEVICEINFO_UNKNOWN,
        DeviceInfo.DEVICEINFO_UNKNOWN, 1946939392L);
    assertEquals(YearClass.CLASS_2013, yearClass);
  }

  private YearClass getYearClass(int numCores, int maxFreqKHz, long memoryBytes) {
    mockStatic(DeviceInfo.class);
    when(DeviceInfo.getNumberOfCPUCores()).thenReturn(numCores);
    when(DeviceInfo.getCPUMaxFreqKHz()).thenReturn(maxFreqKHz);
    when(DeviceInfo.getTotalMemory((Context) anyObject())).thenReturn(memoryBytes);
    YearClass yearClass = YearClass.get(null);
    PowerMockito.verifyStatic();

    return yearClass;
  }

  @Test
  public void testBuildingFromInt() {
    assertEquals(YearClass.CLASS_2008, YearClass.fromIntValue(2008));
    assertEquals(YearClass.CLASS_2009, YearClass.fromIntValue(2009));
    assertEquals(YearClass.CLASS_2010, YearClass.fromIntValue(2010));
    assertEquals(YearClass.CLASS_2011, YearClass.fromIntValue(2011));
    assertEquals(YearClass.CLASS_2012, YearClass.fromIntValue(2012));
    assertEquals(YearClass.CLASS_2013, YearClass.fromIntValue(2013));
    assertEquals(YearClass.CLASS_2014, YearClass.fromIntValue(2014));
  }
}

yearclass-sample/src/main/java/com/facebook/device/yearclass/sample/MainActivity.java
/* This file provided by Facebook is for non-commercial testing and evaluation
 * purposes only.  Facebook reserves all rights not expressly granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * FACEBOOK BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.facebook.device.yearclass.sample;

import android.app.Activity;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.StrictMode;
import android.widget.TextView;
import com.facebook.device.yearclass.YearClass;

public class MainActivity extends Activity {
  private static final String PREF_FILE = "YearClass";
  private static final String PREF_NAME = "yearclass";
  private TextView mYearClass;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    if (BuildConfig.DEBUG) {
      StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
              .detectAll()
              .penaltyLog()
              .build());
      StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
              .detectAll()
              .penaltyLog()
              .penaltyDeath()
              .build());
    }

    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    GetOrComputeYearClass findYearClass = new GetOrComputeYearClass();
    findYearClass.execute();
    mYearClass = (TextView) findViewById(R.id.year_class);
  }

  private class GetOrComputeYearClass extends AsyncTask<Void, Void, Integer> {

    @Override
    protected Integer doInBackground(Void... voids) {
      int yearClass = YearClass.CLASS_UNKNOWN;
      SharedPreferences prefs = getSharedPreferences(PREF_FILE, 0);
      if (prefs.contains(PREF_NAME)) {
        yearClass = prefs.getInt(PREF_NAME, YearClass.CLASS_UNKNOWN);
      }
      //Try again if device was previously unknown.
      if (yearClass == YearClass.CLASS_UNKNOWN) {
        yearClass = YearClass.get(getApplicationContext());
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(PREF_NAME, yearClass);
        editor.apply();
      }
      return yearClass;
    }

    @Override
    protected void onPostExecute(Integer yearClass) {
      //update UI
      mYearClass.setText(Integer.toString(yearClass));
    }
  }
}


yearclass/src/main/java/com/facebook/device/yearclass/YearClass.java
/* Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.facebook.device.yearclass;

import android.content.Context;

import java.util.ArrayList;
import java.util.Collections;

public class YearClass {
  // Year definitions
  public static final int CLASS_UNKNOWN = -1;
  public static final int CLASS_2008 = 2008;
  public static final int CLASS_2009 = 2009;
  public static final int CLASS_2010 = 2010;
  public static final int CLASS_2011 = 2011;
  public static final int CLASS_2012 = 2012;
  public static final int CLASS_2013 = 2013;
  public static final int CLASS_2014 = 2014;
  public static final int CLASS_2015 = 2015;
  public static final int CLASS_2016 = 2016;

  private static final long MB = 1024 * 1024;
  private static final int MHZ_IN_KHZ = 1000;

  private volatile static Integer mYearCategory;

  private static final String UNKNOWN = "UNKNOWN";
  private static final String SINGLE_CORE = "SINGLE CORE";
  private static final String DUAL_CORE = "DUAL CORE";
  private static final String TRIPLE_CORE = "TRIPLE CORE";
  private static final String QUAD_CORE = "QUAD CORE";

  /**
   * Entry Point of YearClass. Extracts YearClass variable with memoizing.
   * Example usage:
   * <p>
   * <pre>
   *   int yearClass = YearClass.get(context);
   * </pre>
   */
  public static int get(Context c) {
    if (mYearCategory == null) {
      synchronized(YearClass.class) {
        if (mYearCategory == null) {
          mYearCategory = categorizeByYear2016Method(c);
        }
      }
    }
    return mYearCategory;
  }

  private static void conditionallyAdd(ArrayList<Integer> list, int value) {
    if (value != CLASS_UNKNOWN) {
      list.add(value);
    }
  }

  /**
   * This formulation of year class smooths out the distribution of devices in the field
   * in early 2016 so that the buckets are a bit more even in size and performance metrics
   * (specifically app startup time, scrolling perf, animations) are more uniform within
   * the buckets than with the 2014 calculations.
   */
  private static int categorizeByYear2016Method(Context c) {
    long totalRam = DeviceInfo.getTotalMemory(c);
    if (totalRam == DeviceInfo.DEVICEINFO_UNKNOWN) {
      return categorizeByYear2014Method(c);
    }

    if (totalRam <= 768 * MB) {
      return DeviceInfo.getNumberOfCPUCores() <= 1 ? CLASS_2009 : CLASS_2010;
    }
    if (totalRam <= 1024 * MB) {
      return DeviceInfo.getCPUMaxFreqKHz() < 1300 * MHZ_IN_KHZ ? CLASS_2011 : CLASS_2012;
    }
    if (totalRam <= 1536 * MB) {
      return DeviceInfo.getCPUMaxFreqKHz() < 1800 * MHZ_IN_KHZ ? CLASS_2012 : CLASS_2013;
    }
    if (totalRam <= 2048 * MB) {
      return CLASS_2013;
    }
    if (totalRam <= 3 * 1024 * MB) {
      return CLASS_2014;
    }
    return totalRam <= 5 * 1024 * MB ? CLASS_2015 : CLASS_2016;
  }

  /**
   * Calculates the "best-in-class year" of the device. This represents the top-end or flagship
   * devices of that year, not the actual release year of the phone. For example, the Galaxy Duos
   * S was released in 2012, but its specs are very similar to the Galaxy S that was released in
   * 2010 as a then top-of-the-line phone, so it is a 2010 device.
   *
   * @return The year when this device would have been considered top-of-the-line.
   */
  private static int categorizeByYear2014Method(Context c) {
    ArrayList<Integer> componentYears = new ArrayList<Integer>();
    conditionallyAdd(componentYears, getNumCoresYear());
    conditionallyAdd(componentYears, getClockSpeedYear());
    conditionallyAdd(componentYears, getRamYear(c));
    if (componentYears.isEmpty())
      return CLASS_UNKNOWN;
    Collections.sort(componentYears);
    if ((componentYears.size() & 0x01) == 1) {  // Odd number; pluck the median.
      return componentYears.get(componentYears.size() / 2);
    } else { // Even number. Average the two "center" values.
      int baseIndex = componentYears.size() / 2 - 1;
      // There's an implicit rounding down in here; 2011.5 becomes 2011.
      return componentYears.get(baseIndex) +
          (componentYears.get(baseIndex + 1) - componentYears.get(baseIndex)) / 2;
    }
  }

  /**
   * Calculates the year class by the number of processor cores the phone has.
   * Evaluations are based off the table below:
   * <table border="1">
   * <thead>
   * <tr><th width="50%">Amount</th><th>Year</th></tr>
   * <thead>
   * <tbody>
   * <tr><td>>4 or More</td><td>2012</td></tr>
   * <tr><td>2 or 3</td><td>2011</td></tr>
   * <tr><td>1</td><td>2008</td></tr>
   * </tbody>
   * </table>
   *
   * @return the year in which top-of-the-line phones had the same number of processors as this phone.
   */
  private static int getNumCoresYear() {
    int cores = DeviceInfo.getNumberOfCPUCores();
    if (cores < 1) return CLASS_UNKNOWN;
    if (cores == 1) return CLASS_2008;
    if (cores <= 3) return CLASS_2011;
    return CLASS_2012;
  }

  /**
   * Evaluates the processor's type
   * @return a String which contains the processor's name
   */
  public static String getCoreDescription() {
    int cores = DeviceInfo.getNumberOfCPUCores();
    if (cores < 1) return UNKNOWN;
    if (cores == 1) return SINGLE_CORE;
    if (cores == 2) return DUAL_CORE;
    if (cores == 3) return TRIPLE_CORE;
    return QUAD_CORE;
  }

  /**
   * Calculates the year class by the clock speed of the cores in the phone.
   * Evaluations are based off the table below:
   * <table border="1">
   * <thead>
   * <tr><th width="50%">Amount</th><th>Year</th></tr>
   * <thead>
   * <tbody>
   * <tr><td>>2GHz</td><td>2014</td></tr>
   * <tr><td><=2GHz</td><td>2013</td></tr>
   * <tr><td><=1.5GHz</td><td>2012</td></tr>
   * <tr><td><=1.2GHz</td><td>2011</td></tr>
   * <tr><td><=1GHz</td><td>2010</td></tr>
   * <tr><td><=600MHz</td><td>2009</td></tr>
   * <tr><td><=528MHz</td><td>2008</td></tr>
   * </tbody>
   * </table>
   *
   * @return the year in which top-of-the-line phones had the same clock speed.
   */
  private static int getClockSpeedYear() {
    long clockSpeedKHz = DeviceInfo.getCPUMaxFreqKHz();
    if (clockSpeedKHz == DeviceInfo.DEVICEINFO_UNKNOWN) return CLASS_UNKNOWN;
    // These cut-offs include 20MHz of "slop" because my "1.5GHz" Galaxy S3 reports
    // its clock speed as 1512000. So we add a little slop to keep things nominally correct.
    if (clockSpeedKHz <= 528 * MHZ_IN_KHZ) return CLASS_2008;
    if (clockSpeedKHz <= 620 * MHZ_IN_KHZ) return CLASS_2009;
    if (clockSpeedKHz <= 1020 * MHZ_IN_KHZ) return CLASS_2010;
    if (clockSpeedKHz <= 1220 * MHZ_IN_KHZ) return CLASS_2011;
    if (clockSpeedKHz <= 1520 * MHZ_IN_KHZ) return CLASS_2012;
    if (clockSpeedKHz <= 2020 * MHZ_IN_KHZ) return CLASS_2013;
    return CLASS_2014;
  }

  /**
   * Calculates the year class by the amount of RAM the phone has.
   * Evaluations are based off the table below:
   * <table border="1">
   * <thead>
   * <tr><th width="50%">Amount</th><th>Year</th></tr>
   * <thead>
   * <tbody>
   * <tr><td>>2GB</td><td>2014</td></tr>
   * <tr><td><=2GB</td><td>2013</td></tr>
   * <tr><td><=1.5GB</td><td>2012</td></tr>
   * <tr><td><=1GB</td><td>2011</td></tr>
   * <tr><td><=512MB</td><td>2010</td></tr>
   * <tr><td><=256MB</td><td>2009</td></tr>
   * <tr><td><=128MB</td><td>2008</td></tr>
   * </tbody>
   * </table>
   *
   * @return the year in which top-of-the-line phones had the same amount of RAM as this phone.
   */
  private static int getRamYear(Context c) {
    long totalRam = DeviceInfo.getTotalMemory(c);
    if (totalRam <= 0) return CLASS_UNKNOWN;
    if (totalRam <= 192 * MB) return CLASS_2008;
    if (totalRam <= 290 * MB) return CLASS_2009;
    if (totalRam <= 512 * MB) return CLASS_2010;
    if (totalRam <= 1024 * MB) return CLASS_2011;
    if (totalRam <= 1536 * MB) return CLASS_2012;
    if (totalRam <= 2048 * MB) return CLASS_2013;
    return CLASS_2014;
  }
}
